### ISSUE TITLE
faster WeightedRandomSampler implementation based on alias method

### ISSUE THREAD

**Issue by guo-yong-zhi** [pytorch/pytorch#94131]
### 🚀 The feature, motivation and pitch

Since the weights are discrete and often fixed, I think the `WeightedRandomSampler` with replacement could be based on [alias method](https://en.wikipedia.org/wiki/Alias_method). In this implementation, the random values can be drawn from the distribution in O(1) time. The weight list has the same size as the dataset. When it comes to large datasets, the new method can be better.
Here is my solution, I implemented a `WeightedRandomSampler2` with scipy:
```python
from torch.utils.data import WeightedRandomSampler
from scipy.stats.sampling import DiscreteAliasUrn

class WeightedRandomSampler2(WeightedRandomSampler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        urng = np.random.default_rng()
        self.rng = DiscreteAliasUrn(self.weights, random_state=urng)
        
    def __iter__(self):
        if self.replacement:
            yield from iter(self.rng.rvs(self.num_samples))
        else:
            return super().__iter__()
```

### Alternatives

https://www.keithschwarz.com/darts-dice-coins/
https://github.com/Tecnarca/Vose-Alias-Method
https://gist.github.com/orlp/e9b31d3397a7dd3e34d6bc862ce3b88d

### Additional context

```python
import numpy as np
w = np.random.random(10_000_000)
w /= w.sum()
```

```python
%%time
s1 = WeightedRandomSampler(w, 3)
```
>Wall time: 0 ns

```python
%%time
list(s1)
```
>Wall time: 63.1 ms
[729931, 4124692, 4903810]

```python
%%time
s2 = WeightedRandomSampler2(w, 3)
```
>Wall time: 235 ms

```python
%%time
list(s2)
```
>Wall time: 0 ns
[3887197, 7289214, 2292607]

We can see that although the initialization of `WeightedRandomSampler2` is slower, its random number generation is much faster. Typically, the initialization runs only once, while the generation runs tens of thousands of times. 
Furthermore, if given a larger weight list, say `w = np.random.random(100_000_000)`, the `WeightedRandomSampler` will raise a `RuntimeError: number of categories cannot exceed 2^24`, but the number generation of the `WeightedRandomSampler2` still tasks `Wall time: 0 ns`.
Therefore, I propose to implement the sampler using alias method.

**Comment by ejguan** [comment]
I don't think this is an apple-to-apple comparison. `WeightedRandomSampler` would initialize random indices when `__iter__` is called for the sake of reseeding per epoch. And it should take only `O(1)` time to yield each index. But `WeightedRandomSampler2` would initialize the `DiscreteAliasUrn` in `__init__`.

To make the benchmarking more solid, can you move `self.rng = DiscreteAliasUrn(self.weights, random_state=urng)` to `__iter__` function? Then, you can do benchmark it using:
```py
%%time
s1 = WeightedRandomSampler(w, 3)
it1 = iter(s1)
```
```py
%%time
list(s1)
```

**Comment by ejguan** [comment]
> Is reseeding matter? If so, I think we could replace the `rng` instead of re-initializing `DiscreteAliasUrn` when `__iter__` is called.

It definitely matters because users want to have different order of data per epoch. And, in order to control determinism, users normally want to seed sampler per epoch with a new seed.

I have a noob question. For `WeightedRandomSampler2`. are the results same  for each iteration?

**Comment by guo-yong-zhi** [comment]
>For WeightedRandomSampler2. are the results same for each iteration?

No. They are random and different for each iteration.
And I don't think calling  `random.seed` can make the result more random.

**Comment by ejguan** [comment]
The expected behavior for PY users is:
```py
for epoch in range(10):
    torch.manual_seed(epoch)
    for d in dataloader:
        ...
```
We need to support setting different seeds per epoch. It means the `DiscreteAliasUrn` has to be reset when new `__iter__` is called.

**Comment by guo-yong-zhi** [comment]
I guess that can be done by passing pytorch's gobal random generator or something alike to `DiscreteAliasUrn.__init__`. DiscreteAliasUrn will be plan B if we can pre-initialize something of the original WeightedRandomSampler.

**Comment by ejguan** [comment]
So, what would be the performance result with:
```py
class WeightedRandomSampler2(WeightedRandomSampler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        
    def __iter__(self):
        urng = np.random.default_rng()  # You can replace it with torch.
        self.rng = DiscreteAliasUrn(self.weights, random_state=urng)
        if self.replacement:
            yield from iter(self.rng.rvs(self.num_samples))
        else:
            return super().__iter__()
```

**Comment by guo-yong-zhi** [comment]
Slower than original WeightedRandomSampler
